doc/developer/design/20210413_source_sink_resource_sharing.md

```diff
@@ -0,0 +1,269 @@
+             /                        \
+            /                           \
+      kafka_src/1                     kafka_src/2
+          |                                |
+        view1                            view2
+```
+
+### Today's workaround
+
+If you know the complete set of columns you care about ahead of time, you can
+coax an efficient implementation out of Materialize today with an intermediate
+materialized view:
+
+```sql
+CREATE SOURCE kafka_src FROM KAFKA BROKER '...' TOPIC 'top' FORMAT AVRO ... ;
+CREATE MATERIALIZED VIEW kafka_view AS SELECT col1, col2 FROM kafka_src;
+CREATE VIEW view1 AS SELECT col1 FROM kafka_view;
+CREATE VIEW view2 AS SELECT col2 FROM kafka_view;
+```
+
+This creates only one librdkafka consumer by ensuring that `kafka_src` is
+only instantiated once.
+
+```
+       kafka_src/1
+  +---------------------+
+  | librdkafka consumer |
+  +---------------------+
+  |    Avro decoding    |
+  +---------------------+
+            |
+        kafka_view
+         /      \
+      view1    view2
+```
+
+The downside, of course, is that you must commit ahead of time to the columns
+you care about. What if, down the road, you realize you need a `view3` that
+extracts `col3`? You have two unappealing options:
+
+  1. Delete `kafka_view`, `view1`, and `view2`. Recreate `kafka_view` with
+    `col3` as well, then create `view1`, `view2`, and `view3` atop the new
+    `kafka_view`.
+
+     This is both annoying to type and will interrupt the availability of
+     `view1` and `view2`.
+
+  2. Create another instantiation of `kafka_src` by materializing `view3`
+     atop `kafka_src` directly.
+
+     This is ergonomic but results in a suboptimal steady state.
+
+### Dynamic resource sharing
+
+With dynamic resource sharing, you get all the ergonomics of creating views
+as you please, but a system that converges on the same efficience that you'd
+get if you'd pre-committed to those views. The key insight is that we can
+create a temporary backfill consumer when a new instance is instantiated:
+
+```
+          steady-state primary                    temporary backfill
+         +---------------------+                 +---------------------+
+         | librdkafka consumer |                 | librdkafka consumer |
+         +---------------------+                 +---------------------+
+         |    Avro decoding    |                 |    Avro decoding    |
+         +---------------------+                 +---------------------+
+       /                        \                          |
+      /                           \                        |
+kafka_src/1                     kafka_src/2           kafka_src/3
+    |                                |                     |
+  view1                            view2                 view3
+```
+
+Once that backfill consumer has caught up to the steady-state primary
+consumer—i.e., once the consumer offsets are identical— we can cut `kafka_src/3`
```

Comment:
The semantics of the offsets are not quite clear for me so maybe this point is wrong.

If we think of a source as a differential dataflow collection (which is what it produces in the end) I would expect the cut-over logic to be based on timestamps and not offsets. So as the two collections advance in time, once we have reached the same timestamp we can do the switch over at the differential dataflow level.

Reply:
I can imagine a bunch of consistency errors, where the backfill timestamps are "real time" and are beyond those timestamps that will eventually be cut over to. I think mechanically this could be made to work, but it comes with some non-trivial logic.